api4/user_test.go

```diff
@@ -5882,6 +5889,78 @@ func TestThreadSocketEvents(t *testing.T) {
 		require.Truef(t, caught, "User should have received %s event", model.WebsocketEventThreadReadChanged)
 	})
 
+	_, resp, err = th.Client.UpdateThreadReadForUser(th.BasicUser.Id, th.BasicTeam.Id, rpost.Id, rpost.CreateAt)
+	require.NoError(t, err)
+	CheckOKStatus(t, resp)
+
+	t.Run("Listen for read event 2", func(t *testing.T) {
+		var caught bool
+		func() {
+			for {
+				select {
+				case ev := <-userWSClient.EventChannel:
+					if ev.EventType() == model.WebsocketEventThreadReadChanged {
+						caught = true
+
+						data := ev.GetData()
+						require.EqualValues(t, rpost.CreateAt, data["timestamp"])
+						require.EqualValues(t, float64(0), data["previous_unread_replies"])
+						require.EqualValues(t, float64(0), data["previous_unread_mentions"])
+						require.EqualValues(t, float64(1), data["unread_replies"])
+						require.EqualValues(t, float64(1), data["unread_mentions"])
+
+					}
+				case <-time.After(1 * time.Second):
+					return
+				}
+			}
+		}()
+
+		require.Truef(t, caught, "User should have received %s event", model.WebsocketEventThreadReadChanged)
+	})
+
+	// make sure to follow the thread
+	resp, err = th.Client.UpdateThreadFollowForUser(th.BasicUser.Id, th.BasicTeam.Id, rpost.Id, true)
+	require.NoError(t, err)
+	CheckOKStatus(t, resp)
+
+	// read the thread
+	_, resp, err = th.Client.UpdateThreadReadForUser(th.BasicUser.Id, th.BasicTeam.Id, rpost.Id, replyPost.CreateAt+1)
+	require.NoError(t, err)
+	CheckOKStatus(t, resp)
+
+	// post a reply on the thread
+	_, resp, err = th.Client.CreatePost(&model.Post{ChannelId: th.BasicChannel.Id, Message: "testReply @" + th.BasicUser.Username, UserId: th.BasicUser2.Id, RootId: rpost.Id})
+	require.NoError(t, err)
+	CheckCreatedStatus(t, resp)
+
+	t.Run("Listen for thread updated event after create post", func(t *testing.T) {
+		var caught bool
+		func() {
+			for {
+				select {
+				case ev := <-userWSClient.EventChannel:
+					if ev.EventType() == model.WebsocketEventThreadUpdated {
+						caught = true
+						data := ev.GetData()
+						var thread model.ThreadResponse
+						jsonErr := json.Unmarshal([]byte(data["thread"].(string)), &thread)
+						require.NoError(t, jsonErr)
+
+						require.Equal(t, float64(0), data["previous_unread_replies"])
+						require.Equal(t, float64(0), data["previous_unread_mentions"])
+						require.Equal(t, int64(0), thread.UnreadReplies)
+						require.Equal(t, int64(0), thread.UnreadMentions)
```

Comment:
@agnivade, @ashishbhate I'm cannot test this properly. I post a new reply as BasicUser2 and should get `unread_replies == 1`, but it seems it is always changing the `thread.LastViewedAt` for `BasicUser` to a value slightly larger than `reply.CreateAt`, so it never has `unreads`.
Could we be changing the `LastViewedAt` of the thread upon CreatePost? 
OR any other ideas on why this is happening?

Reply:
> Could we be changing the LastViewedAt of the thread upon CreatePost?

Yes, that seems to be what's happening. This line apparently does that in app/notification.go

```go
opts := store.ThreadMembershipOpts{
		Following:             true,
		IncrementMentions:     incrementMentions,
		UpdateFollowing:       updateFollowing,
		UpdateViewedTimestamp: userID == post.UserId, // for the OP, the last viewed timestamp gets updated
		UpdateParticipants:    userID == post.UserId,
	}
	threadMembership, err := a.Srv().Store.Thread().MaintainMembership(userID, post.RootId, opts)
```

I love these tests and the fact that this is catching things like this!

Reply:
I also verified it manually, and the db entry does get updated whenever I reply to a thread.

Reply:
oh, totally missed that line!
OK it seems that I can't use `client.CreatePost` to create a post as another user...

Reply:
@agnivade, @ashishbhate  I've updated the tests and added a broken test case with some comments to explain the issues.

https://github.com/mattermost/mattermost-server/blob/9aad9602f81e90b8c5bb1a3f5319477fe62a7a2e/api4/user_test.go#L5935-L5946

Reply:
I tried to fix these 2 issues, the tests pass but can you re-review @agnivade, @ashishbhate?